src/Halogen.purs

```diff
@@ -1,159 +1,78 @@
+import Halogen.Effects (HalogenEffects())
+import Halogen.HTML.Renderer.VirtualDOM (renderHTML)
+import Halogen.Internal.VirtualDOM (VTree(), createElement, diff, patch)
+
+type DriverState s =
+  Maybe { node :: HTMLElement
+        , vtree :: VTree
+        , state :: s
+        }
+
+type Driver f eff = forall i. f i -> Eff (HalogenEffects eff) i
+
+runUI :: forall eff s f. Component s f (Eff (HalogenEffects eff)) Void
+      -> s
+      -> Eff (HalogenEffects eff) { node :: HTMLElement, driver :: Driver f eff }
+runUI c s = case renderComponent c s of
+    Tuple html s' -> do
+      ref <- newRef Nothing
+      let vtree = renderHTML (driver ref) html
+          node = createElement vtree
+      writeRef ref $ Just { node: node, vtree: vtree, state: s' }
+      pure { node: node, driver: driver ref }
 
--- | This function provides the low-level implementation of Halogen's DOM update loop.
--- |
--- | The first argument is a function which receives the `Driver` function as an argument and
--- | constructs a `Process` which will update the DOM given an input.
--- |
--- | This function could be reused to create other types of applications based on signal functions
--- | (2D and 3D canvas, text-based, etc.)
-mainLoop :: forall req eff. (Driver req eff -> Eff (HalogenEffects eff) (Tuple HTMLElement (Process req eff))) ->
-                            Eff (HalogenEffects eff) (Tuple HTMLElement (Driver req eff))
-mainLoop buildProcess = do
-  ref <- newRef Nothing
-  go ref
   where
-  go :: Ref (Maybe { process :: Process req eff, node :: HTMLElement }) ->
-        Eff (HalogenEffects eff) (Tuple HTMLElement (Driver req eff))
-  go ref = do
-    Tuple node process <- buildProcess driver
-    writeRef ref $ Just { process: process, node: node }
-    return (Tuple node driver)
-    where
 
-    driver :: Driver req eff
-    driver req = void $ setTimeout globalWindow 0.0 do
-      ms <- readRef ref
-      case ms of
-        Just { process: process, node: node } -> do
-          let work = runSF process (Tuple req node)
-          node' <- head work
-          writeRef ref $ Just { process: tail work, node: node' }
-        Nothing -> log "Error: An attempt to re-render was made during the initial render."
+  driver :: Ref (DriverState s) -> Driver f eff
+  driver ref q = do
+    refVal <- readRef ref
+    case refVal of
+      Nothing -> throwException $ error "Error: An attempt to re-render was made during the initial render."
+      Just { node: node, vtree: prev, state: s } -> do
+        Tuple i s' <- queryComponent c q s
+        case renderComponent c s' of
+          Tuple html s'' -> do
+            let next = renderHTML (driver ref) html
+            node' <- patch (diff prev next) node
+            writeRef ref $ Just { node: node', vtree: next, state: s'' }
+            pure i
+
+actionF :: forall f g. (Functor f, Functor g, Inject f g) => (forall i. i -> f i) -> Free g Unit
```

Comment:
This function and the others could use some documentation. It's not immediately clear what purpose they serve.